include/hipSYCL/sycl/libkernel/spirv/builtins.hpp

```diff
@@ -0,0 +1,475 @@
+    return __spirv_ocl_##name(x);                                              \
+  }
+
+#define HIPSYCL_DEFINE_SPIRV_BUILTIN2(name)                                    \
+  template <class T> HIPSYCL_BUILTIN T __hipsycl_##name(T x, T y) noexcept {   \
+    return __spirv_ocl_##name(x,y);                                              \
+  }
+
+#define HIPSYCL_DEFINE_SPIRV_BUILTIN3(name)                                    \
+  template <class T>                                                           \
+  HIPSYCL_BUILTIN T __hipsycl_##name(T x, T y, T z) noexcept {                 \
+    return __spirv_ocl_##name(x,y,z);                                              \
+  }
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(acos)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(acosh)
+
+
+template<class T>
+HIPSYCL_BUILTIN T __hipsycl_acospi(T x) noexcept {
+  return __hipsycl_acos(x) / M_PI;
+}
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(asin)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(asinh)
+
+template<class T>
+HIPSYCL_BUILTIN T __hipsycl_asinpi(T x) noexcept {
+  return __hipsycl_asin(x) / M_PI;
+}
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(atan)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(atan2)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(atanh)
+
+template<class T>
+HIPSYCL_BUILTIN T __hipsycl_atanpi(T x) noexcept {
+  return __hipsycl_atan(x) / M_PI;
+}
+
+template<class T>
+HIPSYCL_BUILTIN T __hipsycl_atan2pi(T x, T y) noexcept {
+  return __hipsycl_atan2(x, y) / M_PI;
+}
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(cbrt)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(ceil)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(copysign)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(cos)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(cosh)
+
+template<class T>
+HIPSYCL_BUILTIN T __hipsycl_cospi(T x) noexcept {
+  return __hipsycl_cos(x * M_PI);
+}
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(erf)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(erfc)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(exp)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(exp2)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(exp10)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(pow)
+
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(expm1)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(fabs)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(fdim)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(floor)
+HIPSYCL_DEFINE_SPIRV_BUILTIN3(fma)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(fmax)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(fmin)
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(fmod)
+
+// Unsupported
+template<class T>
+T __hipsycl_fract(T x, T* ptr) noexcept;
+
+// Unsupported
+template<class T, class IntPtr>
+T __hipsycl_frexp(T x, IntPtr y) noexcept;
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN2(hypot)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(ilogb)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(lgamma)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(tgamma)
+
+template<class T, class IntPtr>
+HIPSYCL_BUILTIN T __hipsycl_lgamma_r(T x, IntPtr y) noexcept {
+  return __spirv_ocl_lgamma_r(x, y);
+}
+
+HIPSYCL_DEFINE_SPIRV_BUILTIN(log)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(log2)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(log10)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(log1p)
+HIPSYCL_DEFINE_SPIRV_BUILTIN(logb)
+
+template<class T>
+HIPSYCL_BUILTIN T __hipsycl_mad(T x, T y, T z) noexcept {
+  return __hipsycl_fma(x, y, z);
```

Comment:
spirv should have mad? https://github.com/intel/llvm/blob/54ede407edeb93b7e9334d1e725f48bf981b2965/sycl/include/CL/sycl/detail/builtins.hpp#L152

Reply:
I think on SPIR-V we could probably map *all* of the SYCL builtins to SPIRV ones because the SYCL kernel library is based on the OpenCL kernel library. 
I have based the initial reimplementation of the SPIR-V builtins on what was previously mapped to SPIR-V builtins/what other backends map to builtins to get this PR going quicker. We can probably optimize a lot in the future :)